### PR TITLE
Fix broadcast error when eltype is inconsistent with getindex

### LINKED ISSUE #38422: Type assert error in broadcasting on nightly
Running the tests for AutomotiveSimulator errors on nightly (https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/c3bb6df_vs_788b2c7/AutomotiveSimulator.1.6.0-DEV-a896693d33.log) due to a type assert error in broadcasting. 

```jl
feature extraction: Error During Test at /home/kc/PkgEval16/dev/AutomotiveSimulator/test/test_features.jl:119
  Got exception outside of a @test
  TypeError: in typeassert, expected AbstractVector{var"#s825"} where var"#s825"<:(AbstractVector{var"#s825"} where var"#s825"<:Union{Missing, Float64}), got a value of type Vector{Vector{T} where T}
  Stacktrace:
    [1] copy(bc::Base.Broadcast.Broadcasted{Base.Broadcast.DefaultArrayStyle{1}, Tuple{Base.OneTo{Int64}}, AutomotiveSimulator.var"#25#26"{typeof(dist_to_front_neighbor), Roadway{Float64}, Vector{EntityScene{VehicleState, VehicleDef, Int64}}}, Tuple{Vector{Int64}}})
      @ Base.Broadcast ./broadcast.jl:928
    [2] materialize
      @ ./broadcast.jl:881 [inlined]
    [3] broadcast(f::AutomotiveSimulator.var"#25#26"{typeof(dist_to_front_neighbor), Roadway{Float64}, Vector{EntityScene{VehicleState, VehicleDef, Int64}}}, As::Vector{Int64})
      @ Base.Broadcast ./broadcast.jl:819
    [4] extract_feature(ft::SceneFeature, feature::Function, roadway::Roadway{Float64}, scenes::Vector{EntityScene{VehicleState, VehicleDef, Int64}}, ids::Vector{Int64})
      @ AutomotiveSimulator ~/PkgEval16/dev/AutomotiveSimulator/src/feature-extraction/features.jl:96
....
``` 

This type assert comes from:

https://github.com/JuliaLang/julia/blob/3dc54a23ac93109399357c68c9fc998815f25cdc/base/broadcast.jl#L924-L928

and was introduced in https://github.com/JuliaLang/julia/pull/30485.

Is the package doing something wrong or is this type assert a bit too strict?

cc @nalimilan

### PR DESCRIPTION
**Pull request by nalimilan**
In that case we can infer the return type as `Union{}`, which triggers a type assertion error. This used to work in Julia 1.5. (I'm not sure this kind of thing is legitimate but this is probably not the best place to throw an error about it anyway.)

Fixes https://github.com/JuliaLang/julia/issues/38422.

### CODE REVIEW

**Comment by mbauman** [comment]
Wait, so the compiler infers `Union{}` and _doesn't_ completely crap out when it actually gets a value?

**Comment by mbauman** [comment]
Oooh, I see.  It's because we expect the `Number` to use indexing/eltype when guessing the return type, but _don't actually do that_ when evaluating the expression.

File: test/broadcast.jl [hunk 1]
```diff
@@ -990,3 +990,18 @@ p0 = copy(p)
...
                   [[1, 2],[missing, 1]])
 end
+
+@testset "Issue #28382: eltype inconsistent with getindex" begin
+    struct Cyclotomic <: Number
+    end
+
+    Base.getindex(x::Cyclotomic, i::Integer) = i
+
+    Base.length(x::Cyclotomic) = 1
```
Comment (mbauman) [inline test/broadcast.jl:1000]:
```suggestion
```

These lines are broken in a different way and not needed to test this behavior.

File: test/broadcast.jl [hunk 2]
```diff
@@ -990,3 +990,18 @@ p0 = copy(p)
...
+    Base.eltype(::Type{<:Cyclotomic}) = Tuple{Int,Int}
+
+    Base.:*(c::T, x::Cyclotomic) where {T<:Real} = [1, 2]
+    Base.:*(x::Cyclotomic, c::T) where {T<:Real} = [1, 2]
+
+    @test Cyclotomic() .* [2, 3] == [[1, 2], [1, 2]]
+end
```
Comment (simeonschaub) [inline test/broadcast.jl:1007]:
```suggestion
end

```
Looks like a missing trailing newline.
@staticfloat Shouldn't this have been caught by the whitespace buildbot?

Comment (staticfloat) [reply test/broadcast.jl:1007]: No, I think the whitespace buildbot only checks for trailing whitespace.

Comment (simeonschaub) [reply test/broadcast.jl:1007]: Ah, I see, it's marked as a TODO

**Comment by mbauman** [comment]
Maybe we should just ask inference for the `getindex` too since we're already in the business of pestering inference?

**Comment by nalimilan** [comment]
> Maybe we should just ask inference for the `getindex` too since we're already in the business of pestering inference?

Indeed. In general, assuming that `getindex` is inferrable, inference should give the same result as the current system which is more complex. See https://github.com/JuliaLang/julia/pull/39295.

Though I think it would be safer not to make that change in 1.6, so better backport this one, and only merge https://github.com/JuliaLang/julia/pull/39295 to master?

**Comment by nalimilan** [comment]
@mbauman @KristofferC Should we merge and backport this? #39295 seems to be harder to implement, and too disruptive for 1.6 anyway.

**Comment by mbauman** [comment]
Yes, this seems reasonable to me — especially as a patch for 1.6's behavior.

**Pull request merged**